src/awkward/operations/structure.py

```diff
@@ -1460,6 +1460,60 @@ def apply(layout):
         return out
 
 
+def unflatten(array, counts, highlevel=True):
+    """
+    Args:
+        array: Data to create an array with an additional level from.
+        counts (int or array): Number of elements the new level should have.
+        highlevel (bool): If True, return an #ak.Array; otherwise, return
+            a low-level #ak.layout.Content subclass.
+
+    Returns an array with an additional level of nesting. This is roughly the
+    inverse of #ak.flatten, where `counts` were obtained by #ak.num (both with
+    `axis=1`).
+    For example,
+
+        >>> original = ak.Array([[0, 1, 2], [], [3, 4], [5], [6, 7, 8, 9]])
+        >>> counts = ak.num(original)
+        >>> array = ak.flatten(original)
+        >>> counts
+        <Array [3, 0, 2, 1, 4] type='5 * int64'>
+        >>> array
+        <Array [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] type='10 * int64'>
+        >>> ak.unflatten(array, counts)
+        <Array [[0, 1, 2], [], ... [5], [6, 7, 8, 9]] type='5 * var * int64'>
+    """
+    layout = ak.operations.convert.to_layout(
+        array, allow_record=False, allow_other=False
+    )
+    if isinstance(counts, int):
+        out = ak.layout.RegularArray(layout, counts)
```

Comment:
Is there a reason for raising an exception for `counts == 0`? I could imagine users actually wanting to unflatten with `counts=0`, because they have routines that do `flatten` followed by `unflatten` and the input just ended up being in the form of `"0 * 0 * dtype"`.

Reply:
Well, RegularArray can't have a `size` of `0`: this is part of the specification in the [RegularArray documentation](https://awkward-array.readthedocs.io/en/latest/ak.layout.RegularArray.html) (see the `__init__` of the Python pseudocode). It's also prohibited by RegularArray's `is_valid` function:

```python
>>> import awkward as ak
>>> import numpy as np
>>> ak.Array(ak.layout.RegularArray(ak.layout.NumpyArray(np.array([1, 2, 3])), 0), check_valid=True)
ValueError: at layout (RegularArray): size < 1
```

This means that functions using RegularArrays freely assume that `size` is not `0`, so the consequences are unpredictable: it doesn't fail right away (though this validity constraint _could_, since `size` is just a scalar and checking it is _O(1)_...) but some operation on it might fail.

In looking for alternative representations that give you this desired behavior for `counts == 0`, I realized that we will have to support RegularArrays with `size == 0`. They can be made fairly naturally:

```python
>>> array1 = ak.Array([1, 2, 3, 4, 5])
>>> array2 = ak.Array([[1, 2, 3], [], [4, 5]])
>>> array1[np.newaxis][0:0, 0:0]
<Array [] type='0 * 0 * int64'>
>>> array2[np.newaxis][0:0, 0:0]
<Array [] type='0 * 0 * var * int64'>
```

but the second of these is a RegularArray with `size == 0`:

```python
>>> array2[np.newaxis][0:0, 0:0].layout
<RegularArray size="0">
    <content><ListArray64>
        <starts><Index64 i="[]" offset="0" length="0" at="0x000000000000"/></starts>
        <stops><Index64 i="[]" offset="0" length="0" at="0x000000000000"/></stops>
        <content><NumpyArray format="l" shape="5" data="1 2 3 4 5" at="0x5576b495c4b0"/></content>
    </ListArray64></content>
</RegularArray>
```

I'll open an issue to find all the places where this was assumed and relax the constraint. Go ahead and let `counts == 0`; the consequences will be dealt with in that other issue and PR.